EIPS/eip-20-token-standard.md

```diff
@@ -0,0 +1,184 @@
+## Preamble
+
+    EIP: 20
+    Title: ERC-20 Token Standard
+    Author: Fabian Vogelsteller <fabian@ethereum.org>, Vitalik Buterin <vitalik.buterin@ethereum.org>
+    Type: Standard
+    Category: ERC
+    Status: Accepted
+    Created: 2015-11-19
+
+
+## Simple Summary
+
+A standard interface for tokens.
+
+
+## Abstract
+
+The following standard allows for the implementation of a standard API for tokens within smart contracts.
+This standard provides basic functionality to transfer tokens, as well as allow tokens to be approved so they can be spent by another on-chain third party.
+
+
+## Motivation
+
+A standard interface allows any tokens on Ethereum to be re-used by other applications: from wallets to decentralized exchanges.
+
+
+## Specification
+
+## Token
+### Methods
+
+**NOTE**: Callers MUST handle `false` from `returns (bool success)`.  Callers MUST NOT assume that `false` is never returned!
+
+
+#### name
+
+Returns the name of the token - e.g. `"MyToken"`.
+
+OPTIONAL - This method can be used to improve useability,
+but interfaces and other contracts MUST NOT expect these values to be present.
+
+
+``` js
+function name() constant returns (string name)
+```
+
+
+#### symbol
+
+Returns the symbol of the token. E.g. "HIX".
+
+OPTIONAL - This method can be used to improve useability,
+but interfaces and other contracts MUST NOT expect these values to be present.
+
+``` js
+function symbol() constant returns (string symbol)
+```
+
+
+
+#### decimals
+
+Returns the number of decimals the token uses - e.g. `8`, means to divide the token amount by `100000000` to get its user representation.
+
+OPTIONAL - This method can be used to improve useability,
+but interfaces and other contracts MUST NOT expect these values to be present.
+
+``` js
+function decimals() constant returns (uint8 decimals)
+```
+
+
+#### totalSupply
+
+Returns the total token supply.
+
+``` js
+function totalSupply() constant returns (uint256 totalSupply)
+```
+
+
+
+#### balanceOf
+
+Returns the account balance of another account with address `_owner`.
+
+``` js
+function balanceOf(address _owner) constant returns (uint256 balance)
+```
+
+
+
+#### transfer
+
+Transfers `_value` amount of tokens to address `_to`, and MUST fire the `Transfer` event.
+The function SHOULD `throw` if the `_from` account balance does not have enough tokens to spend.
+
+A token contract which creates new tokens SHOULD trigger a Transfer event with the `_from` address set to `0x0` when tokens are created.
```

Comment:
I would add a recommendation to throw if the sender value is 0, as it's most likely an error. If the token wants to have a explicit burn function it should implement so or it should recommend users to send it to any other invalid address.

Reply:
It is fine having explicit burn and mint functions but they should call Transfer events as defined in the standard and monitored on etherscan.

Reply:
Anybody any thoughts on that?
I really want to merge that ASAP

Reply:
@frozeman The token implementations I've seen so far (e.g. [OpenZeppelin's](https://github.com/OpenZeppelin/zeppelin-solidity/blob/aad25205cdc54243ccfe82a1bd35c08904d2a152/contracts/token/BasicToken.sol#L22), but also many others) seem to make no check that `msg.sender` is not `0x00..00`, of `0xff..ff`, or some other magic number.

Personally, I'd leave this out of the standard's text, instead of adding such provisions retroactively, since it was never widely discussed.

Instead of adding another possibly contentious point, that will keep discussion running ad infinitum, just leave it undefined.

Reply:
Removing it is contentious because it's already there and block scanners use it. Whether individual dev teams like it or not, the block scanners need to find the tokens, they find the tokens through transfer events.

Reply:
@Hackdom Nobody is suggesting removing the `Transfer` event from the standard.

Reply:
@Arachnid right, I'm talking about removing the Transfer trigger on minting. It's been in there for a while right? The scanners use the Transfer event to track tokens. We want to remove the `Transfer(0x0,...)` now because some teams didn't like the implementation and went their own way? I say just finalize and let's go.

Reply:
I guess my point is this thing has been pending for a while and leaving it pending like this is not productive. Tokens are huge now and this is being used, adopters need clarity, not all of the details to be perfect. @MicahZoltu was direct about putting discussion in ERC223 a while ago. We need 3-4 core devs to make a decision and finalize like within a week...and be done.

Reply:
I'm still not convinced that this `SHOULD` adds any real value and it wasn't in the original ERC20 so my vote is still to remove it.  _HOWEVER_, I don't think it is important enough to stop this PR from getting merged/approved.

Reply:
There's changing what the spec requires - eg, introducing backwards incompatibilities or new features - and there's adding advice to implementers, or reflecting defacto standardised behaviour. I think recommending a transfer event on minting falls into the latter category rather than the former.

Reply:
@Arachnid See my argument here: https://github.com/ethereum/EIPs/pull/610#issuecomment-321948122
TL;DR: Without this being a **MUST** (which it can't be in this PR), tooling can't leverage it meaningfully without creating bugs elsewhere.

I believe etherscan _tries_ to interpret `transfer from 0` as `mint` and it results in some tokens having incorrect token allocation charts which leads to significant user confusion.  As a **SHOULD**, I think would encourage other applications to do what etherscan has done, thinking that most tokens will follow the SHOULD, when in fact I'm unconvinced that most do.  So we end up with people building applications/interfaces that are wrong some non-trivial portion of the time with no clear mechanism for resolving that.  If this were left out, the theory is that interfaces would stop trying to build against something that isn't widely adopted and therefore they would be less likely to build features that are permanently in a half broken state.

Reply:
@MicahZoltu has the important point which is clarity, defining the standard, and closing. Maybe even just you three @frozeman @Arachnid and @MicahZoltu since you are the most reputable with investment in this discussion, decide and close. I say we all agree now to go with that decision rather than try to agree on the decision itself. Anyone that disagrees should present a valid argument against deciding and closing.

Reply:
I'm just about the least reputable person I know.  I recommend not including me in any list of "reputable persons".  😄 

I'm 👍 on merging this with or without the **SHOULD**.  I just wanted to make sure my arguments against it were clear and well understood before those that make the merge decisions do so.  If @Arachnid / @frozeman (or whoever make the final merge decision) believe they understand my arguments but still disagree with them then consider the issue resolved (unless further discussion/clarification is desired, in which case I'll happily oblige).

Reply:
AFAIR, at the initial submission of #20, the creation (minting) of tokens was not taken into consideration. The proposal aimed to standartise an interface for transfer of tokens. 

In retrospect, I'd say considerations for creation of tokens should have been kept out of this standard proposal, seeing how it has caused a lot of uncertainty.

My comment above:

> EDIT: That is, instead of adding another possibly contentious point that will keep discussion running ad infinitum, just leave it undefined.

... was meant to highlight this fact: creation of tokens shouldn't have been within scope of this standard's text.

It's still not too late to leave a `Minted` event outside the scope of this particular standard.